urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/server/AbstractINDIServer.java

```diff
@@ -0,0 +1,812 @@
+  }
+
+  /**
+   * The thread listens to the server socket and when a client connects, it is
+   * added to the list of clients.
+   */
+  @Override
+  public void run() {
+    try {
+      socket = new ServerSocket(listeningPort);
+    } catch (IOException e) {
+      System.err.println("Could not listen on port: " + listeningPort + " (maybe busy)");
+      return; // The thread will stop
+    }
+
+    System.err.println("Listening to port " + listeningPort);
+    System.err.flush();
+
+    mainThreadRunning = true;
+
+    while (mainThreadRunning) {
+      Socket clientSocket;
+
+      try {
+        clientSocket = socket.accept();
+      } catch (IOException e) {
+//   This is usually the escape point of the thread when the server is stopped.
+        System.err.println("Server has stopped listening to new Client connections");
+        mainThreadRunning = false;
+        return; // The thread will stop
+      }
+
+      if (clientSocket != null) {
+        if (acceptClient(clientSocket)) {
+          INDIClient client = new INDIClient(clientSocket, this);
+
+          clients.add(client);
+
+          connectionWithClientEstablished(client);
+        } else {
+          try {
+            clientSocket.close();
+          } catch (IOException e) {
+          }
+
+          // System.err.println("Client " + clientSocket.getInetAddress() + " rejected");
+        }
+      }
+    }
+  }
+
+  /**
+   * Gets if the server is listening for new Clients to connect.
+   *
+   * @return
+   * <code>true</code> if the server is listening for new Clients.
+   * <code>false</code> otherwise.
+   */
+  protected boolean isServerRunning() {
+    return mainThreadRunning;
+  }
+
+  /**
+   * Stops the server from listening new Clients. All connections with existing
+   * clients are also broken.
+   */
+  protected void stopServer() {
+    mainThreadRunning = false;
+    try {
+      socket.close();  // Close the socket in order to avoid accepting new connections
+    } catch (IOException e) {
+    }
+
+    for (int i = 0 ; i < clients.size() ; i++) {
+      if (clients.get(i) instanceof INDIClient) {
+        ((INDIClient) clients.get(i)).disconnect();
+      }
+    }
+  }
+
+  /**
+   * Loads all INDI for Java Drivers in a JAR file.
+   *
+   * @param jarFileName The jar file from which to load Drivers.
+   * @throws INDIException if there is a problem with the JAR file
+   */
+  public synchronized void loadJavaDriversFromJAR(String jarFileName) throws INDIException {
+    if (isAlreadyLoaded(jarFileName)) {
+      throw new INDIException("JAR file already loaded.");
+    }
+
+    ArrayList<String> list = getClassesInJAR(jarFileName);
+
+    File file = new File(jarFileName);
+    URL url;
+
+    try {
+      url = file.toURI().toURL();
+    } catch (MalformedURLException e) {
+      throw new INDIException("Error reading JAR file for Class Loader.");
+    }
+
+    URL[] urls = new URL[]{url};
+    ClassLoader cl = new URLClassLoader(urls);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *cl*. Execution paths that do not contain closure statements were detected. Ensure *cl* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).